lte/gateway/c/oai/tasks/sgw/pgw_pco.c

```diff
@@ -315,20 +314,60 @@ int pgw_process_pco_link_mtu_request(
   return pgw_pco_push_protocol_or_container_id(pco_resp, &poc_id_resp);
 }
 
-int pgw_process_pco_pcscf_ip_address_req(
-  protocol_configuration_options_t* const pco_resp,
-  const pco_protocol_or_container_id_t* const poc_id)
+//------------------------------------------------------------------------------
+int pgw_process_pco_pcscf_ipv4_address_req(
+    protocol_configuration_options_t* const pco_resp)
 {
+  if (!spgw_config.pgw_config.pcscf.ipv4_addr.s_addr) {
+    OAILOG_ERROR(
+        LOG_SPGW_APP,
+        "PCO_CI_P_CSCF_IPV4_ADDRESS not configured. Ignoring the containerID "
+        "\n");
+    return RETURNok;
+  }
   pco_protocol_or_container_id_t poc_id_resp = {0};
-  //TODO-Uncomment after copying P-CSCF Address from config
-  //in_addr_t pcscf_ipv4_addr;
+  in_addr_t pcscf_ipv4_addr = spgw_config.pgw_config.pcscf.ipv4_addr.s_addr;
+  uint8_t pcscf_ipv4_array[4];
 
   OAILOG_DEBUG(
-    LOG_SPGW_APP,
-    "PCO: Protocol identifier PCO_CI_P_CSCF_IPV4_ADDRESS_REQUEST \n");
-  poc_id_resp.id = PCO_CI_P_CSCF_IPV4_ADDRESS_REQUEST;
-  poc_id_resp.length = 4;
-  // TODO - Copy P-CSCF Address from config
+      LOG_SPGW_APP, "PCO: Protocol identifier PCO_CI_P_CSCF_IPV4_ADDRESS \n");
+  poc_id_resp.id       = PCO_CI_P_CSCF_IPV4_ADDRESS;
+  poc_id_resp.length   = 4;
+  pcscf_ipv4_array[0]  = (uint8_t)(pcscf_ipv4_addr & 0x000000FF);
+  pcscf_ipv4_array[1]  = (uint8_t)((pcscf_ipv4_addr >> 8) & 0x000000FF);
+  pcscf_ipv4_array[2]  = (uint8_t)((pcscf_ipv4_addr >> 16) & 0x000000FF);
+  pcscf_ipv4_array[3]  = (uint8_t)((pcscf_ipv4_addr >> 24) & 0x000000FF);
+  poc_id_resp.contents = blk2bstr(pcscf_ipv4_array, sizeof(pcscf_ipv4_array));
+
+  return pgw_pco_push_protocol_or_container_id(pco_resp, &poc_id_resp);
+}
+
+//------------------------------------------------------------------------------
+int pgw_process_pco_pcscf_ipv6_address_req(
+    protocol_configuration_options_t* const pco_resp)
+{
+  if (!strlen((char*) spgw_config.pgw_config.pcscf.ipv6_addr.s6_addr)) {
+    OAILOG_ERROR(
+        LOG_SPGW_APP,
+        "PCO_CI_P_CSCF_IPV6_ADDRESS not configured. Ignoring the containerID "
+        "\n");
+    // Send P-CSCF IPv4 address if configured
+    if (RETURNok != pgw_process_pco_pcscf_ipv4_address_req(pco_resp)) {
```

Comment:
I am probably missing something, but how is the IPv4 sent?
Is it through pgw_pco_push_protocol_or_container_id?

Reply:
If UE asks for P-CSCF IPv6 address and if it is not configured then we send IPv4 address. pgw_pco_push_protocol_or_container_id() populates the PCO structure with container id, length and contents and returns to the caller handle_s5_create_session_request() in pgw_handlers.c.  Later in sgw_handle_sgi_endpoint_created() PCO is copied to create_session_response_p message and sent to MME APP